### PR TITLE
Dataset collections - initial build-out

### PR DESCRIPTION
**Pull request by ethanteague**
Connects isdapps/adc-code#1588
Relates to #2219
Relates to #2137 

This PR introduces collections support for datasets. 
**Functionality Implemented:**

- Search row support for collections
- Parent Child association on node view (listed links, with see more link to solr search page)
- Collections Faceted search
- Child parent association on node view.

## Note
- Creating search index fields and facets doesn't seem to be (easily) doable, so this mr only has collections fields and facet logic applied to datasets index. Once dkan_sci_metadata is integrated in dkan codebase, we will need to add collections field logic to the primary search index contained within = dkan_sci_search, @dafeder 

## QA Steps
- [x] Visit node/add/dataset and verify presence of Collection autocomplete field.
- [x] Choose 11 published dataset nodes, and edit them by adding the same dataset ref in the collection field to all 11
- [x] Visually verify on child nodes that a singular collection parent link appears on the node view
- [x] Visually verify on the parent node view that a dataset view of three datasets (all from the children ref'd) is listed, with a see more link, below
- [ ] Click the see more link from parent node and visually verify it takes you to a search page with only items from the parent collection, with a collection facet opened and listing the current parent dataset.
- [ ] Search for a child ite's title in the above collection on the search page after clearing all filters.

* Steps below should only be applied when we implement the dkan_sci_index - datasets index has old search view config. For now, these steps are not necessary.
- [ ] Visually verify that the parent node is listed in the view row for the child item
- [ ] Visually verify that datasets that are not associated with a collection do not show a collection label in the view row.

### CODE REVIEW

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+dependencies[] = defaultconfig
+dependencies[] = entityreference
+dependencies[] = facetapi
+dependencies[] = fe_block
+dependencies[] = filter
+dependencies[] = open_data_schema_map
+dependencies[] = strongarm
+dependencies[] = usda_content_types
+dependencies[] = views
+dependencies[] = views_content
```
Comment (dafeder) [inline modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info]: Should require dkan_dataset as well

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<?php
```
Comment (dafeder) [inline modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc]: I wonder if it's a good idea to name something "field_collection" given the specific meaning that has in Drupal? Let's discuss naming.

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.filter.inc [hunk 3]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (dafeder) [inline modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.filter.inc]: I don't think we need this anymore, [DKAN should allow H2](https://github.com/GetDKAN/dkan/blob/7.x-1.x/modules/dkan/dkan_sitewide/dkan_sitewide.features.filter.inc#L31) which was I think the reason this was added?

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.defaultconfig.inc [hunk 4]
```diff
@@ -0,0 +1,116 @@
+<?php
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.defaultconfig.inc]: :+1:

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.fe_block_settings.inc [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<?php
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.fe_block_settings.inc]: :+1:

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc (hunk 2 above)
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc]: :+1:  Renamed to `field_dataset_collections`

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.filter.inc (hunk 3 above)
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.filter.inc]: :+1:

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info (hunk 1 above)
[the remark this answers is not in the document]
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info]: :+1:

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+package = USDA
+dependencies[] = ctools
+dependencies[] = defaultconfig
+dependencies[] = entityreference
+dependencies[] = facetapi
+dependencies[] = fe_block
+dependencies[] = filter
+dependencies[] = open_data_schema_map
+dependencies[] = strongarm
+dependencies[] = usda_content_types
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.info]: :+1:

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc (hunk 2 above)
Comment (dafeder) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc]: @ethanteague I think this isn't quite consistent with other dataset fields. I'd propose `field_dkan_ispartof_ref`

**Comment by dafeder** [comment]
The QA steps don't quite work - the field is there but I don't see any "see more" link or a way to get into a collections-related search page:

image

**Comment by dafeder** [comment]
@ethanteague in addition, this PR failed linting, see CircleCI build

File: modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc (hunk 2 above)
Comment (ethanteague) [reply modules/dkan/dkan_dataset/modules/dkan_dataset_collections/dkan_dataset_collections.features.field_instance.inc]: :+1:

**Comment by ethanteague** [comment]
@dafeder I don't think linting should run on machine generated files, e.g.: features. This will cause every build (with features files) to fail. Our practice on GlobalNet is to not codesniff said files, for this reason. (edited)

If I make the sniff changes, they will be reverted everytime there is a fe, fr, or fu operation on the same file(s). That's gonna be a timesink

**Comment by dafeder** [comment]
@ethanteague I hadn't noticed that all the linting failures were feature files. Weird that features produces ugly code! If you're familiar with a good method for skipping these, do you want to take a shot at updating the linting logic as well? It lives at https://github.com/GetDKAN/dkan/blob/dataset-collections-et/.ahoy/.scripts/dkan-lint.sh

**Comment by kimwdavidson** [comment]
## QA Steps
- [x] Visit node/add/dataset and verify presence of Collection autocomplete field.
**QA PASS**
- [x] Choose 11 published dataset nodes, and edit them by adding the same dataset ref in the collection field to all 11
- [x] Visually verify on child nodes that a singular collection parent link appears on the node view
**QA PASS**
- [x] Visually verify on the parent node view that a dataset view of three datasets (all from the children ref'd) is listed, with a see more link, below
**QA PASS** I saw 10 in the list and a link to "see all" but I think that's what we talked about
- [x] Click the see more link from parent node and visually verify it takes you to a search page with only items from the parent collection, with a collection facet opened and listing the current parent dataset.
**?** So this took me to a search page with the right results, but I don't see a collection facet and I'm not sure where I should see the parent dataset

- [x] Search for a child item's title in the above collection on the search page after clearing all filters.
- [ ] Visually verify that the parent node is listed in the view row for the child item
**QA FAIL** I don't see this in the search results

- [ ] Visually verify that datasets that are not associated with a collection do not show a collection label in the view row.

**Comment by kimwdavidson** [comment]
## QA Steps
- [x] Visit node/add/dataset and verify presence of Collection autocomplete field.
**QA PASS*
- [x] Choose 11 published dataset nodes, and edit them by adding the same dataset ref in the collection field to all 11
- [x] Visually verify on child nodes that a singular collection parent link appears on the node view
**QA PASS**
- [x] Visually verify on the parent node view that a dataset view of three datasets (all from the children ref'd) is listed, with a see more link, below
**QA PASS*
- [x] Click the see more link from parent node and visually verify it takes you to a search page with only items from the parent collection, with a collection facet opened and listing the current parent dataset.
**QA PASS**
- [x] Search for a child item's title in the above collection on the search page after clearing all filters.
- [ ] Visually verify that the parent node is listed in the view row for the child item
*QA FAIL* Still not seeing the parent node listed with the child item in the search results
- [ ] Visually verify that datasets that are not associated with a collection do not show a collection label in the view row.

@ethanteague ^

**Comment by ethanteague** [comment]
My bad @kimwdavidson - those last two steps are not applicable at present, and will be applied when we have the dkan_sci_metadata logic integrated in dkan starter.

**Comment by dafeder** [comment]
@ethanteague although this PR is complete, collections isn't really ready for release until we're reviewed the UX. I'm thinking I'll start making feature branches for these features-in-progress (just like we've done with data dictionaries) and merge these kinds of things into those instead.

**Pull request merged**